patches/tModLoader/Terraria/ModLoader/Core/TmodFile.cs

```diff
@@ -436,4 +445,31 @@ private void Upgrade()
 		Open();
 		// Read contract fulfilled
 	}
+
+	internal bool VerifyHash() => hashVerified ??= _VerifyHash();
+
+	private bool _VerifyHash()
+	{
+		if (hashStartPos == 0)
+			return false;
+
+		using var fs = File.OpenRead(path);
+		fs.Position = hashStartPos;
+		return Hash.SequenceEqual(SHA1.Create().ComputeHash(fs));
+	}
+
+	private void VerifyHashOrThrow(Exception exception)
+	{
+		if (fileStream == null)
+			throw new IOException($"File not open: {path}", exception);
+
+		// It shouldn't be possible to get this far with an unknown hash starting
+		// position:
+		// - erroneous user code would fail due to no file stream or similar,
+		// - saving, etc. does not use this.
+		Debug.Assert(hashStartPos != 0);
+
+		if (!VerifyHash())
+			throw new Exception(Language.GetTextValue("tModLoader.LoadErrorHashMismatchCorrupted"), exception);
```

Comment:
May as well inline this, since the method has no other uses. Then you probably don't need the assert.